src/agent/src/namespace.rs

```diff
@@ -46,74 +31,203 @@ pub fn get_current_thread_ns_path(ns_type: &str) -> String {
     )
 }
 
-// setup_persistent_ns creates persistent namespace without switchin to it.
-// Note, pid namespaces cannot be persisted.
-pub fn setup_persistent_ns(logger: Logger, ns_type: &'static str) -> Result<Namespace, String> {
-    if let Err(err) = fs::create_dir_all(PERSISTENT_NS_DIR) {
-        return Err(err.to_string());
+#[derive(Debug)]
+pub struct Namespace {
+    logger: Logger,
+    pub path: String,
+    persistent_ns_dir: String,
+    ns_type: NamespaceType,
+}
+
+impl<'a> Namespace {
+    pub fn new(logger: &Logger) -> Self {
```

Comment:
Nice! Formal constructors like this avoid lots of interesting bugs. In fact, having had to deal with many such bugs (in go), I wish there was a way to stop the creation of objects without them.

Reply:
Thanks :D